[SILGen] Merge the two SILGen requests

hamishknight opened the pull request:
They both do essentially the same thing, with the single-file request being the single-element case of the whole-module request, with only the call to `getAllForModule` really caring about the difference.

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test macOS

hamishknight commented:
@swift-ci please test macOS

CodaFi commented on lib/SILGen/SILGen.cpp:
```diff
@@ -1817,6 +1817,8 @@ class SILGenModuleRAII {
 
 public:
   void emitSourceFile(SourceFile *sf) {
+    assert(sf->ASTStage == SourceFile::TypeChecked);
```
👍 ✨

CodaFi commented on lib/SILGen/SILGen.cpp:
```diff
@@ -1835,7 +1837,7 @@ class SILGenModuleRAII {
   }
 
-  SILGenModuleRAII(SILModule &M, ModuleDecl *SM) : SGM{M, SM} {}
+  SILGenModuleRAII(SILModule &M) : SGM{M, M.getSwiftModule()} {}
```
`explicit` here - to combat C++ being *bad*

hamishknight replied:
Good catch!

CodaFi commented on lib/SILGen/SILGenRequests.cpp:
```diff
@@ -46,14 +46,18 @@ SourceLoc swift::extractNearestSourceLoc(const SILGenDescriptor &desc) {
+
+  // Whole module doesn't affect the dependency source.
+  if (auto *mod = desc.context.dyn_cast<ModuleDecl *>()) {
+    return {e.getActiveDependencySourceOrNull(), e.getActiveSourceScope()};
```
You can also explicitly return `{nullptr, active-scope}`, but I suppose this is more conservatively correct.

hamishknight replied:
`nullptr` probably makes it a little more clear, I'll switch over to that for now.

CodaFi reviewed (approved):
Only some nits. I wouldn't stop you from merging this.

hamishknight committed:
[SILGen] Assert the SourceFile has been type-checked

hamishknight committed:
[SILGen] Merge the two SILGen requests

They both do essentially the same thing, with the
single-file request being the single-element case
of the whole-module request, with only the call to
`getAllForModule` really caring about the
difference.

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please smoke test

hamishknight commented:
@swift-ci please test Windows
